File: js/src/cli/add.ts
/* eslint-disable no-console */
import chalk from "chalk";
import { Command } from "commander";
import inquirer from "inquirer";
import open from "open";
import { z } from "zod";
import { Composio } from "../sdk";
import { GetConnectorInfoResDTO, GetConnectorListResDTO } from "../sdk/client";
import { ZAuthMode } from "../sdk/types/integration";

type TInputField = {
  name: string;
  displayName?: string;
  display_name?: string;
  expected_from_customer?: boolean;
  required?: boolean;
  message_name?: string;
  type?: string;
};

type THandleActionOptions = {
  force?: boolean;
  skipDefaultConnector?: boolean;
  noBrowser?: boolean;
  integrationId?: string;
  authMode?: string;
  scope?: string[];
  label?: string[];
};

type TAuthScheme = {
  auth_mode: string;
  fields: TInputField[];
};

export default class AddCommand {
  private program: Command;

  constructor(program: Command) {
    this.program = program;
    this.program
      .command("add")
      .description("Add a new app")
      .argument("<app-name>", "The name of the app")
      .option("-f, --force", "Force the connection setup")
      .option(
        "--skip-default-connector",
        "Skip the default connector auth prompt"
      )
      .option("-n, --no-browser", "Don't open browser for verifying connection")
      .option(
        "-i, --integration-id <id>",
        "Specify integration ID to use existing integration"
      )
      .option("-a, --auth-mode <mode>", "Specify auth mode for given app")
      .option(
        "-s, --scope <scope>",
        "Specify scopes for the connection",
        (value, previous: string[]) => previous.concat([value]),
        []
      )
      .option(
        "-l, --label <label>",
        "Labels for connected account",
        (value, previous: string[]) => previous.concat([value]),
        []
      )
      .action(this.handleAction.bind(this));
  }

  private async handleAction(
    appName: string,
    options: THandleActionOptions
  ): Promise<void> {
    const composioClient = new Composio({});
    let integration:
      | GetConnectorInfoResDTO
      | GetConnectorListResDTO
      | undefined;

    if (options.integrationId) {
      integration = await composioClient.integrations.get({
        integrationId: options.integrationId,
      });
    } else {
      integration = await composioClient.integrations.list({
        appName: appName.toLowerCase(),
      });
    }

    let firstIntegration: GetConnectorInfoResDTO | undefined;
    if (
      (integration as GetConnectorListResDTO)?.items?.length === 0 ||
      options.force ||
      options.skipDefaultConnector
    ) {
      const integrationResult = await this.createIntegration(
        appName,
        options.skipDefaultConnector,
        options.authMode,
        options
      );

      if (integrationResult) {
        firstIntegration =
          integrationResult as unknown as GetConnectorInfoResDTO;
      }
    } else {
      firstIntegration = integration as GetConnectorInfoResDTO;
    }
    if (!firstIntegration) {
      console.log(chalk.red("No integration found or created"));
      return;
    }

    const connection = await composioClient.connectedAccounts.list({
      integrationId: firstIntegration.id,
    });

    if (connection.items.length > 0 && !options.force) {
      await this.shouldForceConnectionSetup();
    }

    if (firstIntegration && firstIntegration.id) {
      await this.setupConnections(firstIntegration.id, options);
    } else {
      console.log(chalk.red("Integration ID is undefined"));
    }
  }

  async shouldForceConnectionSetup() {
    const prompt = inquirer.createPromptModule();
    const { shouldForce } = await prompt([
      {
        type: "confirm",
        name: "shouldForce",
        message:
          "A connection already exists. Do you want to force a new connection?",
        default: false,
      },
    ]);

    if (!shouldForce) {
      console.log(
        chalk.yellow("Operation cancelled. Existing connection will be used.")
      );
      process.exit(0);
    }
  }

  private async waitUntilConnected(
    connectedAccountId: string,
    timeout: number = 30000
  ): Promise<void> {
    const composioClient = new Composio({});
    const startTime = Date.now();
    const pollInterval = 3000; // 3 seconds

    while (Date.now() - startTime < timeout) {
      try {
        const data = (await composioClient.connectedAccounts.get({
          connectedAccountId: connectedAccountId,
        })) as { status: string };

        if (data.status === "ACTIVE") {
          return;
        }
      } catch (error) {
        console.error("Error checking connection status:", error);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(
      `Connection did not become active within ${timeout / 1000} seconds`
    );
  }

  private async setupConnections(
    integrationId: string,
    options: Record<string, unknown>
  ): Promise<void> {
    const composioClient = new Composio({});
    const data = await composioClient.integrations.get({ integrationId });
    const { expectedInputFields } = data!;

    const config = await this.collectInputFields(
      expectedInputFields as unknown as TInputField[],
      true
    );

    if (options.scope) {
      config.scopes = (options.scope as string[]).join(",");
    }

    const connectionData = await composioClient.connectedAccounts.initiate({
      integrationId,
      connectionParams: config,
      labels: options.label as string[],
    });

    if (connectionData.connectionStatus === "ACTIVE") {
      console.log(chalk.green("Connection created successfully"));
    }

    if (connectionData.redirectUrl && !options.noBrowser) {
      console.log(
        chalk.white("Redirecting to the app"),
        chalk.blue(connectionData.redirectUrl)
      );
      open(connectionData.redirectUrl);

      await this.waitUntilConnected(connectionData.connectedAccountId);

      console.log(chalk.green("Connection is active"));
      process.exit(0);
    } else if (connectionData.redirectUrl && options.noBrowser) {
      console.log(
        chalk.white(
          "Please authenticate the app by visiting the following URL:"
        ),
        chalk.blue(connectionData.redirectUrl)
      );
      console.log(
        chalk.green("Waiting for the connection to become active...")
      );

      await this.waitUntilConnected(connectionData.connectedAccountId);

      console.log(chalk.green("Connection is active"));
      process.exit(0);
    }
  }

  private async createIntegration(
    appName: string,
    skipDefaultConnectorAuth: boolean = false,
    userAuthMode?: string,
    options?: THandleActionOptions
  ) {
    const composioClient = new Composio({});
    const app = await composioClient.apps.get({
      appKey: appName.toLowerCase(),
    });

    if (app.no_auth) {
      console.log(
        chalk.green(
          `The app '${appName}' does not require authentication. You can connect it directly.\n`
        )
      );
      process.exit(0);
    }

    const testConnectors = app.testConnectors || [];

    const config: Record<string, unknown> = {};
    let useComposioAuth = true;
    const authSchemeExpectOauth = ["bearer_token", "api_key", "basic"];
    if (
      !app.no_auth &&
      testConnectors.length > 0 &&
      !skipDefaultConnectorAuth &&
      testConnectors.find((connector) => connector.auth_mode === userAuthMode)
    ) {
      const prompt = inquirer.createPromptModule();
      const { doYouWantToUseComposioAuth } = await prompt({
        type: "confirm",
        name: "doYouWantToUseComposioAuth",
        message: "Do you want to use Composio Auth?",
      });
      useComposioAuth = doYouWantToUseComposioAuth;
    }

    if (skipDefaultConnectorAuth) {
      useComposioAuth = false;
    }

    const prompt = inquirer.createPromptModule();
    const { integrationName } = await prompt({
      type: "input",
      name: "integrationName",
      message: "Enter the Integration name",
    });

    if (!integrationName) {
      console.log(chalk.red("Integration name is required"));
      return null;
    }

    config.name = integrationName;
    const authSchema: string | undefined =
      userAuthMode ||
      (app.auth_schemes &&
        (app.auth_schemes[0]?.auth_mode as string | undefined));

    const authModes = (app.auth_schemes || []).reduce(
      (acc, scheme: Record<string, unknown>) => {
        acc[scheme.auth_mode as string] = scheme;
        return acc;
      },
      {}
    );

    if (
      authSchema &&
      typeof authSchema === "string" &&
      !authModes[authSchema]
    ) {
      console.log(
        chalk.red(
          `Invalid value for auth_mode, select from ${Object.keys(authModes)}`
        )
      );
      return null;
    }

    const selectedAuthMode = authSchema || Object.keys(authModes)[0];
    const selectedAuthScheme = authModes[selectedAuthMode];

    if (authSchemeExpectOauth.includes(selectedAuthMode.toLowerCase())) {
      return this.handleBasicAuth(
        app,
        selectedAuthMode,
        selectedAuthScheme as TAuthScheme,
        config,
        integrationName
      );
    }

    return this.handleOAuth(
      app,
      selectedAuthMode,
      selectedAuthScheme as TAuthScheme,
      config,
      integrationName,
      options?.noBrowser ?? false,
      options?.scope ?? [],
      useComposioAuth
    );
  }

  private async handleBasicAuth(
    app: Record<string, unknown>,
    authMode: string,
    authScheme: TAuthScheme,
    config: Record<string, unknown>,
    integrationName: string
  ) {
    const composioClient = new Composio({});
    const authConfig = await this.collectInputFields(authScheme.fields);

    const integration = await composioClient.integrations.create({
      appId: app.appId as string,
      authScheme: authMode as unknown as z.infer<typeof ZAuthMode>,
      useComposioAuth: false,
      name: integrationName,
      authConfig,
    });

    return integration;
  }

  private async handleOAuth(
    app: Record<string, unknown>,
    authMode: string,
    authScheme: TAuthScheme,
    config: Record<string, unknown>,
    integrationName: string,
    noBrowser: boolean,
    scopes: string[],
    useComposioAuth: boolean
  ) {
    if (useComposioAuth) {
      return this.setupIntegration(
        app as {
          appId: string;
        },
        authMode,
        useComposioAuth,
        {},
        integrationName
      );
    }

    const authConfig = await this.collectInputFields(
      authScheme.fields as TInputField[]
    );

    if (scopes) {
      authConfig.scopes = scopes.join(",");
    }

    return this.setupIntegration(
      app as {
        appId: string;
      },
      authMode,
      useComposioAuth,
      authConfig,
      integrationName
    );
  }

  async collectInputFields(
    fields: TInputField[],
    isConnection = false
  ): Promise<Record<string, unknown>> {
    const config: Record<string, unknown> = {};

    for (const field of fields) {
      if (field.expected_from_customer && !isConnection) {
        continue;
      }

      const prompt = inquirer.createPromptModule();
      const { [field.name]: value } = await prompt({
        type: "input",
        name: field.name,
        message: (field.displayName || field.display_name) as string,
      });

      if (value) {
        config[field.name] = value;
      }
    }

    return config;
  }

  async setupIntegration(
    app: {
      appId: string;
    },
    authMode: string,
    useComposioAuth: boolean,
    config: Record<string, unknown>,
    name: string
  ) {
    const composioClient = new Composio({});
    const integration = await composioClient.integrations.create({
      appId: app.appId,
      authScheme: authMode as unknown as z.infer<typeof ZAuthMode>,
      useComposioAuth,
      name,
      authConfig: config,
    });
    return integration;
  }
}

File: python/composio/cli/add.py
"""
Add new integration.

Usage:
    composio add [options]
"""

import typing as t
import webbrowser

import click

from composio.cli.context import Context, ensure_login, pass_context
from composio.cli.utils.decorators import handle_exceptions, pass_entity_id
from composio.cli.utils.helpfulcmd import HelpfulCmd
from composio.client import Composio, Entity
from composio.client.collections import (
    AppAuthScheme,
    AppModel,
    AuthSchemeField,
    AuthSchemeType,
    IntegrationModel,
)
from composio.client.exceptions import ComposioClientError
from composio.constants import DEFAULT_ENTITY_ID
from composio.utils.url import get_web_url


class AddIntegrationExamples(HelpfulCmd):
    examples = [
        click.style("composio add <app_name>", fg="green")
        + click.style("                      # Add a new integration\n", fg="black"),
        click.style("composio add <app_name> --no-browser", fg="green")
        + click.style(
            "         # Add a new integration without opening the browser\n", fg="black"
        ),
        click.style("composio add <app_name> -i <integration_id>", fg="green")
        + click.style(
            "  # Add a new integration using an existing integration ID\n", fg="black"
        ),
    ]


@click.command(name="add", cls=AddIntegrationExamples)
@click.help_option("--help", "-h", "-help")
@click.argument("name", type=str)
@click.option(
    "--no-browser",
    is_flag=True,
    default=False,
    help="Don't open browser for verifying connection",
)
@click.option(
    "-i",
    "--integration-id",
    type=str,
    help="Specify integration ID to use existing integration",
)
@click.option(
    "-a",
    "--auth-mode",
    type=str.upper,
    help="Specify auth mode for given app.",
    metavar="MODE",
)
@click.option(
    "-s",
    "--scope",
    "scopes",
    type=str,
    help="Specify scopes for the connection.",
    multiple=True,
)
@click.option(
    "--force",
    is_flag=True,
    help="Override the existing account.",
)
@click.option(
    "-l",
    "--label",
    "labels",
    help="Labels for connected account.",
    multiple=True,
)
@pass_entity_id
@handle_exceptions()
@ensure_login
@pass_context
def _add(
    context: Context,
    name: str,
    scopes: t.Tuple[str, ...],
    entity_id: str,
    integration_id: t.Optional[str],
    labels: t.List[str],
    no_browser: bool = False,
    auth_mode: t.Optional[str] = None,
    force: bool = False,
) -> None:
    """Add a new integration."""
    add_integration(
        name=name.lower().strip(),
        context=context,
        entity_id=entity_id,
        integration_id=integration_id,
        no_browser=no_browser,
        auth_mode=auth_mode,
        scopes=scopes,
        force=force,
        labels=list(labels or []),
    )


def _replace_connection() -> bool:
    """Prompt user to check if they want to replace the connection or not."""
    return (
        click.prompt(
            "> Do you want to replace the existing connection?",
            type=click.Choice(
                choices=("y", "n"),
                case_sensitive=False,
            ),
        )
        == "y"
    )


def _collect_input_fields(fields: t.List[AuthSchemeField]) -> t.Dict:
    """Collect"""
    inputs = {}
    for _field in fields:
        field = _field.model_dump()
        if field.get("required", False):
            value = input(f"> Enter {field.get('display_name', field.get('name'))}: ")
            if not value:
                raise click.ClickException(
                    f"{field.get('display_name', field.get('name'))} is required"
                )
        else:
            value = input(
                f"> Enter {field.get('display_name', field.get('name'))} (Optional):"
            ) or t.cast(
                str,
                field.get("default"),
            )
        inputs[field.get("name")] = value
    return inputs


def _load_integration(
    context: Context,
    integration_id: t.Optional[str] = None,
) -> t.Optional[IntegrationModel]:
    """Load integration model."""
    if integration_id is None:
        return None

    for integration in context.client.integrations.get():
        if integration.id == integration_id:
            return integration

    raise click.ClickException(f"No integration found with ID: `{integration_id}`")


def add_integration(
    name: str,
    context: Context,
    entity_id: str = DEFAULT_ENTITY_ID,
    integration_id: t.Optional[str] = None,
    no_browser: bool = False,
    auth_mode: t.Optional[str] = None,
    scopes: t.Optional[t.Tuple[str, ...]] = None,
    labels: t.Optional[t.List] = None,
    force: bool = False,
) -> None:
    """
    Add integration.

    :param name: App name.
    :param context: CLI runtime context.
    :param entity_id: Entity ID to use for creating integration.
    :param no_browser: Don't open browser.
    :param auth_mode: Preferred auth mode.
    :param scopes: List of scopes for the connected account.
    """
    entity = context.client.get_entity(id=entity_id)
    integration = _load_integration(
        context=context,
        integration_id=integration_id,
    )

    try:
        existing_connection = entity.get_connection(app=name)
    except ComposioClientError:
        existing_connection = None

    if existing_connection is not None and not force:
        context.console.print(
            f"[yellow]Warning: An existing connection for {name} was found.[/yellow]\n"
        )
        if not _replace_connection():
            context.console.print(
                "\n[green]Existing connection retained. No new connection added.[/green]\n"
            )
            return None

    if existing_connection is not None and force:
        context.console.print(
            f"[yellow]Warning: Replacing existing connection for {name}.[/yellow]\n"
        )

    context.console.print(
        f"\n[green]> Adding integration: {name.capitalize()}...[/green]\n"
    )
    app = t.cast(AppModel, context.client.apps.get(name=name))
    if app.no_auth:
        raise click.ClickException(f"{app.name} does not require authentication")

    auth_schemes = app.auth_schemes or []
    if len(auth_schemes) == 0:
        context.console.print(f"{app.name} does not need authentication")
        return None

    auth_modes = {auth_scheme.auth_mode: auth_scheme for auth_scheme in auth_schemes}
    if auth_mode is not None and auth_mode not in auth_modes:
        raise click.ClickException(
            f"Invalid value for `auth_mode`, select from `{set(auth_modes)}`"
        )

    if auth_mode is not None:
        auth_mode = t.cast(AuthSchemeType, auth_mode)
        auth_scheme = auth_modes[auth_mode]
    elif len(auth_modes) == 1:
        ((auth_mode, auth_scheme),) = auth_modes.items()
    else:
        auth_mode = t.cast(
            AuthSchemeType,
            click.prompt(
                "Select auth mode: ",
                type=click.Choice(choices=list(auth_modes)),
            ),
        )
        auth_scheme = auth_modes[auth_mode]

    if auth_mode.lower() in ("basic", "api_key", "bearer_token"):
        return _handle_basic_auth(
            entity=entity,
            app_name=name,
            auth_mode=auth_mode,
            auth_scheme=auth_scheme,
            labels=labels,
        )
    return _handle_oauth(
        entity=entity,
        client=context.client,
        app_name=name,
        auth_scheme=auth_scheme,
        no_browser=no_browser,
        integration=integration,
        scopes=scopes,
        use_composio_auth=len(app.testConnectors or []) != 0,
        labels=labels,
    )


def _handle_oauth(
    entity: Entity,
    client: Composio,
    app_name: str,
    auth_scheme: AppAuthScheme,
    no_browser: bool = False,
    integration: t.Optional[IntegrationModel] = None,
    scopes: t.Optional[t.Tuple[str, ...]] = None,
    labels: t.Optional[t.List] = None,
    use_composio_auth: bool = False,
) -> None:
    """Handle no auth."""
    auth_config = {}
    if not use_composio_auth:
        auth_config.update(
            _collect_input_fields(
                fields=auth_scheme.fields,
            )
        )

    if scopes is not None:
        if auth_config.get("scopes") is not None:
            scopes = tuple(set([*auth_config["scopes"].split(","), *scopes]))
        auth_config["scopes"] = ",".join(scopes)

    connection = entity.initiate_connection(
        app_name=app_name.lower(),
        redirect_url=get_web_url(path="redirect"),
        integration=integration,
        auth_mode="OAUTH2",
        auth_config=auth_config,
        use_composio_auth=use_composio_auth,
        force_new_integration=len(scopes or []) > 0,
        labels=labels,
    )
    if not no_browser:
        webbrowser.open(
            url=str(connection.redirectUrl),
        )
    click.echo(
        f"Please authenticate {app_name} in the browser and come back here. "
        f"URL: {connection.redirectUrl}"
    )
    click.echo(f"⚠ Waiting for {app_name} authentication...")
    connection.wait_until_active(client=client)
    click.echo(
        f"✔ {app_name} added successfully with ID: {connection.connectedAccountId}"
    )


def _handle_basic_auth(
    entity: Entity,
    app_name: str,
    auth_mode: str,
    auth_scheme: AppAuthScheme,
    integration: t.Optional[IntegrationModel] = None,
    labels: t.Optional[t.List] = None,
) -> None:
    """Handle basic auth."""
    auth_config = _collect_input_fields(
        fields=auth_scheme.fields,
    )
    connection = entity.initiate_connection(
        app_name=app_name.lower(),
        auth_mode=auth_mode,
        auth_config=auth_config,
        integration=integration,
        use_composio_auth=False,
        force_new_integration=True,
        labels=labels,
    )
    click.echo(
        f"✔ {app_name} added successfully with ID: {connection.connectedAccountId}"
    )
